internal/gcs/bridge.go

```diff
@@ -0,0 +1,356 @@
+package gcs
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/Microsoft/hcsshim/internal/guid"
+	"github.com/sirupsen/logrus"
+)
+
+type requestMessage interface {
+	SetActivityID(id guid.GUID)
+}
+
+type responseMessage interface {
+	Err() error
+}
+
+// rpc represents an outstanding rpc request to the guest
+type rpc struct {
+	proc rpcProc
+	id   int64
+	req  requestMessage
+	resp responseMessage
+	err  error
+	ch   chan struct{}
+}
+
+// bridge represents a communcations bridge with the guest. It handles the
+// transport layer but (mostly) does not parse or construct the message payload.
+type bridge struct {
+	Timeout time.Duration
+	mu      sync.Mutex
+	nextID  int64
+	rpcs    map[int64]*rpc
+	conn    io.ReadWriteCloser
+	rpcCh   chan *rpc
+	notify  notifyFunc
+	closed  bool
+	log     *logrus.Entry
+}
+
+var (
+	errBridgeClosed = errors.New("bridge closed")
+
+	// Any non-async RPC must respond in this time or the bridge
+	// connection will be terminated.
+	bridgeFailureTimeout = time.Minute * 5
+)
+
+type notifyFunc func(*containerNotification) error
+
+// newBridge returns a bridge on the given connection. It calls notify when a
+// notification message arrives from the guest. It logs transport errors and
+// traces using the given logrus entry.
+func newBridge(conn io.ReadWriteCloser, notify notifyFunc, log *logrus.Entry) *bridge {
+	return &bridge{
+		conn:    conn,
+		rpcs:    make(map[int64]*rpc),
+		rpcCh:   make(chan *rpc),
+		notify:  notify,
+		log:     log,
+		Timeout: bridgeFailureTimeout,
```

Comment:
I wonder if we want to handle this as a Maximum rather than an initial. IE: The caller of newBridge could just overwrite this value by setting `b.Timeout = time.Hour * 1`. If the intention is to have this be the max then `b.Timeout` is the user timeout `<= bridgeFailureTimeout` right?

Reply:
I think the caller might want to set 1 hour if they are debugging the guest, though.

Reply:
Eh. That's fine for now. I guess we just need to make sure the caller appropriately sets the timeout. It would be my future intention that with profiling we minimize these timeouts to the 100% success case but shortest failure timeouts. Hopefully this can become say 30seconds or something one day

Reply:
Maybe for LCOW... WCOW is going to need a longer timeout for creating the container...

Reply:
Hmm. Lets leave it for now.

I am wondering if the caller should fully control this then. It seems to make sense that if "not set" the behavior is to wait for ever but then the impl of each CreateContainer, Exec, etc, can choose the appropriate timeout for what they know to be the case. Given we control both timeouts internally it seems just easy to make sure we set appropriate ones as it does to rely on a 5 minute timeout. Although this does make your debugging example harder.

Reply:
Yes, we could add a per-call timeout to RPC, and we could have a way for debug mode to turn these off globally.